### PR TITLE
Fix mmap failure handling in MemoryMappedFile

### PR DESCRIPTION
**Pull request by ChiahongHong**
This PR fixes `MemoryMappedFile::open()` to check `mmap()` failures correctly.

```
$ man mmap

RETURN VALUES
     Upon successful completion, mmap() returns a pointer to the mapped region.  Otherwise, a value of
     MAP_FAILED is returned and errno is set to indicate the error.
```

`mmap()` returns `MAP_FAILED` (`((void *)-1)`) on failure, not `nullptr`. The previous check could treat a failed mapping as a successful open, leaving `data_` with an invalid pointer `0xffffffffffffffff`.

I also added two tests to verify the handling of empty files and failure cases.

### CODE REVIEW

File: Source/Engine/MemoryMappedFile.cpp [hunk 1]
```diff
@@ -68,10 +68,11 @@ bool MemoryMappedFile::open(const char* path) {
   length_ = static_cast<size_t>(sb.st_size);
 
   data_ = mmap(nullptr, length_, PROT_READ, MAP_SHARED, fd_, 0);
-  if (data_ == nullptr) {
+  if (data_ == MAP_FAILED) {
```
[the remark this answers is not in the document]
Comment (ChiahongHong) [reply Source/Engine/MemoryMappedFile.cpp:71]: Treating `nullptr` as failure here would conflate a theoretical valid mapping value with the class’ closed sentinel. If we want to harden the class independently, we should instead make `fd_ == -1` the ownership sentinel for `close()`, but I don’t think an extra `data_ == nullptr` failure check is required for this fix.

File: Source/Engine/MemoryMappedFileTest.cpp [hunk 2]
```diff
@@ -124,4 +124,43 @@ TEST(MemoryMappedFileTest, BasicFunctionalities) {
   EXPECT_EQ(mf5.data(), nullptr);
 }
 
+TEST(MemoryMappedFileTest, EmptyFile) {
+  std::filesystem::path tmp_file_path =
+      std::filesystem::temp_directory_path() /
+      "org.openvanilla.mcbopomofo.memorymappedfiletest-empty-file";
```
[the remark this answers is not in the document]
Comment (ChiahongHong) [reply Source/Engine/MemoryMappedFileTest.cpp:130]: Agreed that fixed filenames can theoretically race in parallel runs. However, checking for a randomized non-existing path is still not "atomic"; it only lowers the collision probability. A robust fix should use `mkstemp()` or `mkdtemp()`.

I’m ok with following the reviewers’ preference, either using the current `BasicFunctionalities` approach or switching to `mkstemp()` or `mkdtemp()`.

File: Source/Engine/MemoryMappedFile.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukhnos) [reply Source/Engine/MemoryMappedFile.cpp:71]: I agree with your assessment. Let's not worry about such a theoretical scenario for this PR.

File: Source/Engine/MemoryMappedFileTest.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukhnos) [reply Source/Engine/MemoryMappedFileTest.cpp:130]: Let's not sweat on it for this PR. If we want to fix this, we should as well fix all other such uses in this file.

**Review by lukhnos — APPROVED** [review]
LGTM.

While there's the saving grace that `fstat` gets called first, it's embarrassing that we didn't check for the actual error value. Thank you so much for the fix!
